[GTK][WPE] Add support for showing notification icon

carlosgcampos opened the pull request:
#### abb92629ba7294891e342d8a00304d0bd005fff7
<pre>
[GTK][WPE] Add support for showing notification icon
<a href="https://bugs.webkit.org/show_bug.cgi?id=243786">https://bugs.webkit.org/show_bug.cgi?id=243786</a>

Reviewed by Michael Catanzaro.

Add NotificationResourcesLoader to load the notification resources
following the spec (<a href="https://notifications.spec.whatwg.org/#resources).">https://notifications.spec.whatwg.org/#resources).</a>
For now only the icon image is loaded since it&apos;s the only one actually
used. Resources are passed to the UI process as an extra parameter to
ShowNotification message. This patch only adds suport for notification
icon in the default notifications implementation, but it should also be
exposed in the public API in a follow up.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::show):
(WebCore::Notification::close):
(WebCore::Notification::stop):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/NotificationResources.h: Added.
(WebCore::NotificationResources::create):
(WebCore::NotificationResources::setIcon):
(WebCore::NotificationResources::icon const):
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp: Added.
(WebCore::NotificationResourcesLoader::NotificationResourcesLoader):
(WebCore::NotificationResourcesLoader::resourceIsSupportedInPlatform):
(WebCore::NotificationResourcesLoader::start):
(WebCore::NotificationResourcesLoader::stop):
(WebCore::NotificationResourcesLoader::didFinishLoadingResource):
(WebCore::NotificationResourcesLoader::ResourceLoader::ResourceLoader):
(WebCore::NotificationResourcesLoader::ResourceLoader::~ResourceLoader):
(WebCore::NotificationResourcesLoader::ResourceLoader::cancel):
(WebCore::NotificationResourcesLoader::ResourceLoader::didReceiveResponse):
(WebCore::NotificationResourcesLoader::ResourceLoader::didReceiveData):
(WebCore::NotificationResourcesLoader::ResourceLoader::didFinishLoading):
(WebCore::NotificationResourcesLoader::ResourceLoader::didFail):
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;WebCore::NotificationResources&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::NotificationResources&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/UIProcess/API/APINotificationProvider.h:
(API::NotificationProvider::show):
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:
(WebKitNotificationProvider::show):
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.h:
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp:
(WebKit::WebNotificationManagerProxy::show):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.cpp:
(WebKit::WebNotificationProvider::show):
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.h:
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
(WebKit::IconCache::IconCache):
(WebKit::IconCache::~IconCache):
(WebKit::IconCache::iconPath):
(WebKit::IconCache::iconBytes):
(WebKit::IconCache::unuseIcon):
(WebKit::IconCache::removeUnusedIcons):
(WebKit::IconCache::timerFired):
(WebKit::iconCache):
(WebKit::NotificationService::showNotification):
(WebKit::NotificationService::didCloseNotification):
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showNotification):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::showServiceWorkerNotification):
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::show):

Canonical link: <a href="https://commits.webkit.org/253460@main">https://commits.webkit.org/253460@main</a>
</pre>

mcatanzaro commented on Source/WebCore/Modules/notifications/Notification.cpp:
```diff
@@ -174,14 +176,24 @@ void Notification::show(CompletionHandler<void()>&& callback)
-    if (client->show(*this, scope.release()))
-        m_state = Showing;
+    // Wait for any fetches to complete and notification’s image resource, icon resource, and badge resource to be set (if any),
+    // as well as the icon resources for the notification’s actions (if any).
```
Wow, you've switched to curly apostrophes instead of straight apostrophes. Fancy. :P Are you manually copy/pasting these? It's fine... I'm just amused that you're using a non-typeable character, that's all. There is one other GNOME developer who does this, and I've just never understood the purpose.

carlosgcampos replied:
I just copy pasted from the spec.

mcatanzaro commented on Source/WebCore/Modules/notifications/Notification.h:
```diff
@@ -101,6 +103,7 @@ class Notification final : public ThreadSafeRefCounted<Notification>, public Act
     ScriptExecutionContext* scriptExecutionContext() const final { return ActiveDOMObject::scriptExecutionContext(); }
 
     WEBCORE_EXPORT NotificationData data() const;
+    const RefPtr<NotificationResources>& resources() const { return m_resources; }
```
Returning a `const RefPtr<NotificationResources>&` is slightly weird? This should probably either (a) return a raw pointer `const NotificationResources*` and let the caller decide whether to take a ref, or (b) just return a RefPtr by value so that the caller always receives a ref. Right?

I guess it's personal preference, but returning a RefPtr by reference just seems a little odd to me, since it doesn't modify the refcount.

carlosgcampos replied:
I agree, the intention was to return by value indeed.

mcatanzaro commented on Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp:
```diff
@@ -0,0 +1,157 @@
+    }
+}
+
+void NotificationResourcesLoader::ResourceLoader::didFinishLoading(ResourceLoaderIdentifier, const NetworkLoadMetrics&)
```
You don't need to chain up, or do something with the NetworkLoadMetrics? Aren't there consequences to not reporting things along to the NetworkLoadMetrics? Maybe it won't appear properly in the web inspector?

carlosgcampos replied:
No, the load is shown in the inspector with all the metrics.

mcatanzaro commented on Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp:
```diff
@@ -0,0 +1,157 @@
+
+NotificationResourcesLoader::ResourceLoader::~ResourceLoader()
+{
+}
```
Better to declare the destructor with `= default` in the header if you have nothing to do here.

carlosgcampos replied:
It's done this way to avoid having to include BitmapImage.h in the header, because it's needed by RefPtr. This way we can forward declare it instead.

mcatanzaro commented on Source/WebCore/Modules/notifications/Notification.cpp:
```diff
@@ -174,14 +176,24 @@ void Notification::show(CompletionHandler<void()>&& callback)
+    m_resourcesLoader = makeUnique<NotificationResourcesLoader>(*this);
+    m_resourcesLoader->start([this, client, callback = scope.release()](RefPtr<NotificationResources>&& resources) mutable {
+        CompletionHandlerCallingScope scope { WTFMove(callback) };
+        m_resources = WTFMove(resources);
```
Hmmm, I wonder: what happens if the resource load does not complete before notification is destroyed? I guess:

 * Notification destructor is called
 * Member variables are destroyed in reverse order of declaration
 * The resource loader will execute the callback when it is destroyed (does it?)
 * Callback writes to `m_resources` here... not ideal

I'm not actually sure how it should work. WebKit doesn't use cancellables like GObject code does. I guess it's more normal to use a WeakPtr and return early from the callback if the weakThis is nullptr... but you can't do that here, because Notification owns m_resourceLoader. Hm...

carlosgcampos replied:
I assumed stop or close were called before the destructor, which calls NotificationResourcesLoader::stop(), but I'm not sure that's the case. I'll fix this, thanks!

mcatanzaro commented on Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp:
```diff
@@ -0,0 +1,157 @@
+    m_resources = nullptr;
+
+    for (auto& loader : m_loaders)
+        loader->cancel();
```
What happens to `NotificationResourcesLoader`'s `m_completionHandler` in this case? You have to make sure it gets called before `NotificationResourceLoader` is destroyed, right? But I don't see where that happens if the resource load does not finish.

carlosgcampos replied:
The loader cancel makes didFail callback to be called (synchronously) which calls the completion handler with nullptr.

mcatanzaro commented on Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp:
```diff
@@ -0,0 +1,157 @@
+void NotificationResourcesLoader::ResourceLoader::cancel()
+{
+    m_loader->cancel();
+}
```
Similar question here: what happens to the `NotificationResourcesLoader::ResourceLoader`'s completion handler? Where does that get executed? Does `didFail` get called by the parent class? If so, fine.

carlosgcampos replied:
Same here, all loaders are cancelled and the completion handler is called because the loaders map is empty.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -45,6 +52,173 @@ namespace WebKit {
+                WTF::switchOn(it.value.second,
+                    [](const CString& path) {
+                        if (!path.isNull())
+                            unlink(path.data());
```
I like to check for errors whenever possible, since it makes debugging so much easier when things go unexpectedly wrong:

```suggestion
                        if (!path.isNull())
                            if (unlink(path.data()) == -1)
                                WTFLogAlways("Failed to remove cached icon %s: %s", path.data(), safeStrerror(errno).data());
```

carlosgcampos replied:
This is very unlikely to fail, but ok.

mcatanzaro commented on Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
```diff
@@ -45,6 +52,173 @@ namespace WebKit {
+            return;
+
+        if (!it->value.first)
+            return;
```
In this case, deletion is already pending, right? It took me too long to figure this out, so it's probably worth adding a quick comment to explain.

carlosgcampos replied:
No, this means it's already deleted (or scheduled) at this point, I'm not sure it can really happen, maybe it should be an assert, but it's to ensure we don't overflow the unsigned.

mcatanzaro reviewed (approved):
Nice, I've wanted this for a while. Seems this was more difficult than I would have expected. I'm a little surprised we're the first platform to support notification icons, but whatever.

carlosgcampos committed:
[GTK][WPE] Add support for showing notification icon
https://bugs.webkit.org/show_bug.cgi?id=243786

Reviewed by Michael Catanzaro.

Add NotificationResourcesLoader to load the notification resources
following the spec (https://notifications.spec.whatwg.org/#resources).
For now only the icon image is loaded since it's the only one actually
used. Resources are passed to the UI process as an extra parameter to
ShowNotification message. This patch only adds suport for notification
icon in the default notifications implementation, but it should also be
exposed in the public API in a follow up.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::show):
(WebCore::Notification::close):
(WebCore::Notification::stop):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/NotificationResources.h: Added.
(WebCore::NotificationResources::create):
(WebCore::NotificationResources::setIcon):
(WebCore::NotificationResources::icon const):
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.cpp: Added.
(WebCore::NotificationResourcesLoader::NotificationResourcesLoader):
(WebCore::NotificationResourcesLoader::resourceIsSupportedInPlatform):
(WebCore::NotificationResourcesLoader::start):
(WebCore::NotificationResourcesLoader::stop):
(WebCore::NotificationResourcesLoader::didFinishLoadingResource):
(WebCore::NotificationResourcesLoader::ResourceLoader::ResourceLoader):
(WebCore::NotificationResourcesLoader::ResourceLoader::~ResourceLoader):
(WebCore::NotificationResourcesLoader::ResourceLoader::cancel):
(WebCore::NotificationResourcesLoader::ResourceLoader::didReceiveResponse):
(WebCore::NotificationResourcesLoader::ResourceLoader::didReceiveData):
(WebCore::NotificationResourcesLoader::ResourceLoader::didFinishLoading):
(WebCore::NotificationResourcesLoader::ResourceLoader::didFail):
* Source/WebCore/Modules/notifications/NotificationResourcesLoader.h: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.h:
* Source/WebKit/Shared/Notifications/NotificationManagerMessageHandler.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<WebCore::NotificationResources>::encode):
(IPC::ArgumentCoder<WebCore::NotificationResources>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/UIProcess/API/APINotificationProvider.h:
(API::NotificationProvider::show):
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp:
(WebKitNotificationProvider::show):
* Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.h:
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp:
(WebKit::WebNotificationManagerProxy::show):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.h:
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.cpp:
(WebKit::WebNotificationProvider::show):
* Source/WebKit/UIProcess/Notifications/WebNotificationProvider.h:
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp:
(WebKit::IconCache::IconCache):
(WebKit::IconCache::~IconCache):
(WebKit::IconCache::iconPath):
(WebKit::IconCache::iconBytes):
(WebKit::IconCache::unuseIcon):
(WebKit::IconCache::removeUnusedIcons):
(WebKit::IconCache::timerFired):
(WebKit::iconCache):
(WebKit::NotificationService::showNotification):
(WebKit::NotificationService::didCloseNotification):
* Source/WebKit/UIProcess/Notifications/glib/NotificationService.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::showNotification):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::showServiceWorkerNotification):
* Source/WebKit/WebProcess/Notifications/WebNotificationManager.cpp:
(WebKit::WebNotificationManager::show):

Canonical link: https://commits.webkit.org/253460@main
